### PR TITLE
solving #257 #250 #341

### PR DESCRIPTION
**Pull request by pohaha**
we have fixed the problem with food not being placed randomly. and all tests that broe due to it/ It was done on your branch - so you shold at leat see if it all works with you^) gl.

### CODE REVIEW

**Review by swom — COMMENTED** [review]
There is a way to improve the test for #257, probably @Clemcarle can help. Basically just stitch a for loop around lines `1717` up to `1723` and repeat it a few thousand times, collect the `coordinate`s generated inside the loop in a `vector` and then check that those `coordinate`s' `x` and `y` have mean and variance corresponding to the mean and variance of an uniform distribution with the same ranges of `x` and `y`. 
@richelbilderbeek do you think there is another more elegant way to prove that they are coming from a distribution?

File: food.cpp [hunk 1]
```diff
@@ -29,9 +33,23 @@ bool food::is_eaten() const noexcept {
 
 bool operator==(const food& lhs, const food& rhs) noexcept
 {
-    return lhs.get_x() == rhs.get_x()
-            && lhs.get_y() == rhs.get_y()
-            && lhs.get_color() == rhs.get_color();
+    bool is_color_equal = (lhs.get_color() == rhs.get_color());
```
Comment (TheoPannetier) [inline food.cpp:36]:
```suggestion
    bool is_color_equal = lhs.get_color() == rhs.get_color();

```

File: food.cpp [hunk 2]
```diff
@@ -29,9 +33,23 @@ bool food::is_eaten() const noexcept {
 
 bool operator==(const food& lhs, const food& rhs) noexcept
 {
-    return lhs.get_x() == rhs.get_x()
-            && lhs.get_y() == rhs.get_y()
-            && lhs.get_color() == rhs.get_color();
+    bool is_color_equal = (lhs.get_color() == rhs.get_color());
+    return get_x(lhs) == get_x(rhs) &&
+          (get_y(lhs) == get_y(rhs)) &&
```
Comment (TheoPannetier) [inline food.cpp:38]:
```suggestion
          get_y(lhs) == get_y(rhs) &&

```

File: coordinate.cpp [hunk 3]
```diff
@@ -1,4 +1,5 @@
 #include "coordinate.h"
+#include "game.h"
```
Comment (TheoPannetier) [inline coordinate.cpp:2]:
```suggestion
```

File: enemy.cpp [hunk 4]
```diff
@@ -1,4 +1,5 @@
 #include "enemy.h"
+#include "game.h"
```
Comment (TheoPannetier) [inline enemy.cpp:2]:
```suggestion

```

File: food.cpp [hunk 5]
```diff
@@ -3,6 +3,10 @@
 #include <cassert>
 #include <cmath>
 #include <sstream>
+#include <stdio.h>      /* printf, scanf, puts, NULL */
+#include <stdlib.h>     /* srand, rand */
+#include <time.h>
+#include <iostream>
```
Comment (TheoPannetier) [inline food.cpp:9]:
```suggestion
```

Comment (TheoPannetier) [reply food.cpp:9]: I dont't think these libraries are called in the end

File: enemy.cpp (hunk 4 above)
Comment (TheoPannetier) [reply enemy.cpp:2]: superfluous call, I think?

File: food.h [hunk 6]
```diff
@@ -47,6 +49,15 @@ int get_greenness(const food &f) noexcept;
 /// Get the redness (from the color) of the food
 int get_redness(const food &f) noexcept;
 
+
+
```
Comment (TheoPannetier) [inline food.h:53]:
```suggestion
```

File: game_view.cpp [hunk 7]
```diff
@@ -279,10 +279,15 @@ void game_view::draw_player_coords() noexcept
...
-        str_player_coords += "Player " + p.get_ID() + " x = " + std::to_string(trunc(p.get_x()));
-        str_player_coords += "\nPlayer " + p.get_ID() + " y = " + std::to_string(trunc(p.get_y()));
+        str_player_coords += "Player " + p.get_ID() + " x = " + std::to_string(trunc(get_x(p)));
+        str_player_coords += "\nPlayer " + p.get_ID() + " y = " + std::to_string(trunc(get_y(p)));
         str_player_coords += "\n\n";
     }
+    food f = m_game.get_food()[0];
+    str_player_coords += "Food x = " + std::to_string(trunc(get_x(f)));
+    str_player_coords += "y = " + std::to_string(trunc(get_y(f)));
+    str_player_coords += "\n\n";
```
Comment (TheoPannetier) [inline game_view.cpp:289]: Do we want to keep this? I believe it may still be useful for dev for now but it is placed awkwardly.

**Review by TheoPannetier — APPROVED** [review]
Brilliant work, thanks for fixing the random position 👍

**Comment by TheoPannetier** [comment]
I've reviewed the changes you guys made, @swom could you review the parts I changed?

File: game_view.cpp (hunk 7 above)
Comment (swom) [reply game_view.cpp:289]: yeah maybe in another issue

**Review by swom — COMMENTED** [review]
accepted changes suggested

**Comment by swom** [comment]
fastest answer on this repo

**Comment by TheoPannetier** [comment]
@swom I meant the commits I made earlier in this PR 😅 
Are you happy with my solutions for #411?

**Comment by swom** [comment]
> @swom I meant the commits I made earlier in this PR 😅 Are you happy with my solutions for #411?

Ahhhhhgh, too much stuff on a single pull request, will go back and check, where is it?

**Comment by TheoPannetier** [comment]
> Ahhhhhgh, too much stuff on a single pull request

Yeah, sorry I'll switch to another branch earlier next time

> where is it?

Kinda everywhere. Use the file changed feature of the PR 😉

**Comment by swom** [comment]
Ah yeah ok, you did not implement templates for now. Yeah all seems ok. Will not check in detail and if things are not ok, I will take care.

**Comment by TheoPannetier** [comment]
> 
> 
> There is a way to improve the test for #257, probably @Clemcarle can help. Basically just stitch a for loop around lines `1717` up to `1723` and repeat it a few thousand times, collect the `coordinate`s generated inside the loop in a `vector` and then check that those `coordinate`s' `x` and `y` have mean and variance corresponding to the mean and variance of an uniform distribution with the same ranges of `x` and `y`. @richelbilderbeek do you think there is another more elegant way to prove that they are coming from a distribution?

Just saw this. We actually discussed this with @richelbilderbeek in 20918be, for #250 .
The thing is the test should remain as simple as possible, and the issue doesn't specify that new positions should be sampled in a distribution, just that they should be different. So we stick to the concise and very simple test.

**Comment by swom** [comment]
> > There is a way to improve the test for #257, probably @Clemcarle can help. Basically just stitch a for loop around lines `1717` up to `1723` and repeat it a few thousand times, collect the `coordinate`s generated inside the loop in a `vector` and then check that those `coordinate`s' `x` and `y` have mean and variance corresponding to the mean and variance of an uniform distribution with the same ranges of `x` and `y`. @richelbilderbeek do you think there is another more elegant way to prove that they are coming from a distribution?
> 
> Just saw this. We actually discussed this with @richelbilderbeek in [20918be](https://github.com/tresinformal/game/commit/20918bed3a666ade0baae8b514ba09ed9c1728b6), for #250 . The thing is the test should remain as simple as possible, and the issue doesn't specify that new positions should be sampled in a distribution, just that they should be different. So we stick to the concise and very simple test.

I actually think the original test went as I re-suggested, I remember this conversation. Good. All done then 👍

**Comment by TheoPannetier** [comment]
> 
> 
> Ah yeah ok, you did not implement templates for now. Yeah all seems ok. Will not check in detail and if things are not ok, I will take care.

I did actually use the templated functions in most cases, except where a call to a member function was necessary 😇

**Pull request merged**